File: test-example.py
#!/usr/bin/env python
# This script has been tested with Python 2.7 and 3.4

# Using Selenium + ChromeDriver,
# see https://sites.google.com/a/chromium.org/chromedriver/getting-started
# and http://selenium-python.readthedocs.org/en/latest/installation.html

import os
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC

# Make sure that the current directory is the root of the repo, so that
# --load-extension will succeed.
os.chdir(os.path.join(os.path.dirname(__file__)))

options = webdriver.ChromeOptions()
# Load helper extensions.
options.add_argument('load-extension=helper,helper-app,test-extension')
# Alternatively, you can zip or package every extension and use
# Note: For packaging into a .crx file, you need the private key (key.pem).
# options.add_extension('helper.zip')
# options.add_extension('helper-app.zip')
# options.add_extension('test-extension.zip')

# If you want to know what would happen without the helper extensions, use:
# options.add_argument('load-extension=test-extension')
# (spoiler: The test will fail)


# Note: chromedriver must be available within your PATH.
# If it is not available, use
# driver = webdriver.Chrome(executable_path="/path/to/chromedriver", options)
driver = webdriver.Chrome(chrome_options=options)

print('Visiting page with inline installation button')
# Visit the page with inline installation
driver.get('https://robwu.nl/s/chrome.webstore.install-demo.html')

text_after_install = 'The extension has been installed.'

try:
    # Wait until button is visible
    # http://selenium-python.readthedocs.org/en/latest/waits.html
    button = WebDriverWait(driver, 3).until(
        EC.visibility_of_element_located((By.ID, 'install-extension')))
    assert button.text != text_after_install, \
        'The extension should be disabled before running the test'
    print('Triggering inline install.')
    button.click()

    print('Waiting until completion of installation.')
    # Check whether the extension is installed, by verifying that the
    # button's text equals |text_after_install|.
    locator = (By.ID, 'install-extension')
    button = WebDriverWait(driver, 10).until(
        lambda driver: button.text == text_after_install)
    print('Button found, test passed!')
finally:
    driver.quit()

File: sample/TestExample.java
import java.io.File;
import java.lang.AssertionError;
import java.io.IOException;
import org.openqa.selenium.By;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;


/**
Get ChromeDriver: https://sites.google.com/a/chromium.org/chromedriver/getting-started
Get Selenium    : wget http://selenium-release.storage.googleapis.com/2.44/selenium-server-standalone-2.44.0.jar -O selenium-server-standalone.jar

Start the server:
java -jar selenium-server-standalone.jar

Compile this file:
javac -cp selenium-server-standalone.jar test_example.java

Run this file:
java -cp .:selenium-server-standalone.jar TestExample

If ChromeDriver is not in your PATH, or if you are using Windows, see http://stackoverflow.com/a/17014883.

 */
class TestExample {
    private WebDriver driver;

    TestExample() {
    }

    public void run() {
        File root = getRepoRoot();
        if (root == null) {
            System.err.println("Cannot find sample/ directory. Please run the code from the root of the repository.");
            return;
        }
        String load_extension_arg;
        try {
            // Load the helper extension, app and sample extension.
            String helper_ext = new File(root, "helper").getCanonicalPath();
            String helper_app = new File(root, "helper-app").getCanonicalPath();
            String sample_ext = new File(root, "sample/sample_extension").getCanonicalPath();
            load_extension_arg = String.format("load-extension=%s,%s,%s", helper_ext, helper_app, sample_ext);
            System.out.println(load_extension_arg);
        } catch (IOException e) {
            System.err.println("One of the required directories did not exist: " + e.getMessage());
            return;
        }

        ChromeOptions options = new ChromeOptions();
        options.addArguments(load_extension_arg);

        driver = (WebDriver)new ChromeDriver(options);
        try {
            runTest();
        } finally {
            driver.quit();
        }

    }

    private void runTest() {
        driver.get("https://robwu.nl/s/chrome.webstore.install-demo.html");

        String text_after_install = "The extension has been installed.";

        // Wait until button is visible
        WebElement button = (new WebDriverWait(driver, 3)).until(
                ExpectedConditions.visibilityOfElementLocated(By.id("install-extension")));
        if (button.getText().equals(text_after_install)) {
            throw new AssertionError("The extension should be disabled before running the test");
        }
        System.out.println("Triggering inline install.");
        button.click();

        System.out.println("Waiting until completion of installation.");

        // Check whether the extension is installed, by verifying that the
        // button's text equals |text_after_install|.
        new WebDriverWait(driver, 10).until(
                ExpectedConditions.textToBePresentInElement(button, text_after_install));

        System.out.println("Button found, test passed!");

    }

    private File getRepoRoot() {
        // Get the parent directory of sample = root of repository.
        String SAMPLE_DIR_NAME = "sample";
        File dir = new File(".");
        try {
            dir = dir.getCanonicalFile();
        } catch (IOException e) {
        }
        if (dir.getName().equals(SAMPLE_DIR_NAME)) {
            return dir.getParentFile();
        }
        if (new File(dir, "sample").isDirectory()) {
            return dir;
        }
        return null;
    }

    public static void main(String[] args) {
        new TestExample().run();
    }
}
